File: examples/kafka-jvm/src/clj/pixelcount/spouts/pixel_spout.clj
(ns pixelcount.spouts.pixel_spout
  (:import [storm.kafka SpoutConfig KafkaSpout KafkaConfig KafkaConfig
                        StringScheme ZkHosts]
           [backtype.storm.spout SchemeAsMultiScheme]))


;; Config spelled out below here to get the reader more comfortable with
;; Clojure
;; ^{...} adds metadata to a var definition
(def ^{:doc "Host string for Zookeeper"}
  zk-hosts "streamparse-box:2181"
  )


;; Default path, shouldn't be a need to change this
(def ^{:doc "Zookeeper broker path"}
  zk-broker-path "/brokers")

(def kafka-zk-hosts (ZkHosts. zk-hosts zk-broker-path))

(def ^{:doc "Topic name"}
  topic-name "pixels")

(def ^{:doc "Root path of Zookeeper for spout to store consumer offsets"}
  kafka-zk-root "/kafka_storm")

(def ^{:doc "ID for this Kafka consumer"}
  kafka-consumer-id "pixel_reader")

(def ^{:doc "Kafka spout config definition"}
  spout-config (let [cfg (SpoutConfig. kafka-zk-hosts topic-name kafka-zk-root kafka-consumer-id)]
                  (set! (. cfg scheme) (SchemeAsMultiScheme. (StringScheme.)))
                  ;; During testing, it's usually valuable to force a spout to
                  ;; read from the beginning of a topic
                  (set! (. cfg forceFromStart) true)
                  cfg))

(def spout (KafkaSpout. spout-config))

File: examples/kafka-jvm/src/java/pixelcount/spouts/PixelSpout.java
package pixelcount.spouts;

import storm.kafka.SpoutConfig;
import storm.kafka.KafkaSpout;
import storm.kafka.StringScheme;
import storm.kafka.ZkHosts;
import backtype.storm.spout.SchemeAsMultiScheme;

public class PixelSpout extends KafkaSpout {

	public PixelSpout(SpoutConfig spoutConf) {
		super(spoutConf);
	}

	public PixelSpout() {
		this(PixelSpout.defaultSpoutConfig());
	}

	public static SpoutConfig defaultSpoutConfig() {
		ZkHosts hosts = new ZkHosts("streamparse-box:2181", "/brokers");
		SpoutConfig spoutConf = new SpoutConfig(hosts, "pixels", "/kafka_storm", "pixel_reader");
		spoutConf.scheme = new SchemeAsMultiScheme(new StringScheme());
		spoutConf.forceFromStart = true;
		return spoutConf;
	}
}
